enhancements/shared-and-guaranteed-cpu-pools.md

```diff
@@ -0,0 +1,407 @@
+* Workload 3 - Application user plane pods – L1/L2 function
+  * These are the pods that carry the user plane traffic
+  * These pods will use guaranteed CPUs along with additional isolation for example:
+    * Disable CPU load balancing and CFS quota
+    * Isolate interrupts from the allocated CPUs
+    * Reduce timer ticks (nohz_full)
+  * Generally a small number of pods consuming the bulk of the CPUs on the node
+
+Although there are effectively two CPU pools from an application perspective, Kubernetes only knows about a 
+single pool of CPUs (apart from the reserved CPU pool) and uses this pool for all containers. Kubernetes 
+dynamically allocates CPUs from this single pool for guaranteed QoS containers (using whole CPUs) and 
+updates the CPU affinity of the other containers to ensure they do not run on the guaranteed CPUs.
+
+In order to meet the stringent latency, jitter and performance targets of workload 3, additional kernel 
+tuning is required:
+* Some kernel tuning can be done dynamically and applied at guaranteed container creation time.
+* Unfortunately, some kernel tuning needs to be defined at boot time and is static meaning the same tuning is
+applied to all application CPUs whether it is required or not.
+* Some of these static tunings are problematic for workload 2, such as nohz_full and rcu callbacks. So we
+need the tuning for workload 3 but it is problematic for workload 2.
+
+This proposal provides an option for creating a separate pre-defined pool of CPUs for workload 2 (i.e. 
+shared CPUs) and workload 3 (i.e guaranteed CPUs), which will allow the necessary kernel tuning to be 
+applied to workload 3 without adversely affecting workload 2.
+
+### User Stories
+
+* As a telco service provider, I want to configure the set of CPUs running the L1/L2 functions of a vDU 
+application for extremely low latency, without impacting the other application components running on the 
+same server (on different CPUs).
+
+### Goals
+
+* Allow the configuration of separate pre-defined pools of shared and guaranteed CPUs, with different
+kernel tuning applied to each pool.
+* Automatically assign containers that use whole CPUs and are in a pod with the guaranteed QoS class to the
+guaranteed CPUs.
+* The selection of shared vs. guaranteed CPUs should be transparent to the user - no changes to the pod
+spec should be required.
+* The configuration should be optional and activated with a feature gate.
+* Although the initial usecase for this enhancement is single-node deployments, it should also be usable
+for multi-node deployments.
+
+### Non-Goals
+
+* This enhancement assumes the feature is activated as part of installing the cluster and cannot be
+activated later. However, the sizing of the pools of shared and guaranteed CPUs can be modified (a reboot
+will be required).
+* The use of the shared and guaranteed CPU pools will be activated at the cluster level - it cannot be
```

Comment:
yet, each node can has a slightly different allocation of shared and reserved right?

Reply:
The allocation of shared/reserved/isolated is done in the PerformanceProfile, which is linked to a MachineConfigPool. So to have different allocations of shared/reserved/isolated CPUs you would need to have multiple MachineConfigPools - the nodes in each pool would have the same allocations.

Reply:
Added a note to clarify that nodes can have different shared/guaranteed CPU allocations.